### DISCUSSION TITLE
Cosmetic filter with custom path (url)

### DISCUSSION THREAD

**Discussion by vain0x** [uBlockOrigin/uBlock-issues#1755] [Q&A] [answered]
I want to make a cosmetic filter rule that works only on some path of a site. For example,

```
example.com/foo/*##.bad
```

This rule is invalid. I expected the rule to hide elements with `bad` classes when I'm on a URL that matches `example.com/foo/*`. However, I don't expect it to work when I'm on `example.com/other/` etc. so `example.com##.bad` is not what I want.

Side notes:

- I have read [Static filter syntax · gorhill/uBlock Wiki](https://github.com/gorhill/uBlock/wiki/Static-filter-syntax) and [How to write filters | Adblock Plus Help Center](https://help.eyeo.com/en/adblockplus/how-to-write-filters). Sorry if I just overlook something.
- I have found same question on Reddit:  <https://www.reddit.com/r/uBlockOrigin/comments/6rjszy/element_hiding_for_specific_url_not_domain/>.
    - Answer was pointless.
- version: uBlock Origin 1.38.2, Firefox

Thanks in advance.

**Comment by uBlock-user** [comment, marked as answer]
https://github.com/gorhill/uBlock/wiki/Procedural-cosmetic-filters#subjectmatches-patharg

**Reply by vain0x** [reply]
Thank you!

**Reply by mtxadmin** [reply]
Thank you for the answer, I had the same question in mind.

**Comment by stephenhawk8054** [comment]
@gorhill I see https://github.com/gorhill/uBlock/commit/8b696a691a0871da8200d1806300842d988a4326 . I'm curious of what are the differences of it vs `matches-path`?

**Reply by gorhill** [reply]
`:matches-path` is specific to cosmetic filtering. The new syntax applies to all types of static extended filtering, and the path matching is done before injection time to decide whether a filter should be injected.

The new syntax is equivalent to supporting AdGuard's `[$path=...]` syntax, but I declined to adopt AG's syntax, the way I see it it has [difficult parsing issues](https://github.com/AdguardTeam/tsurlfilter/issues/121) which requires minding [escaping various characters](https://github.com/uBlockOrigin/uAssets/issues/22274#issuecomment-2745408699) (not normally escaped in a literal regex) and I find it too verbose, contributing to cognitive burden for maintainers. I stuck to a syntax which doesn't require any special treatment, there is no need to escape beyond what literal regexes require in JS when using regex form (see [this case](https://github.com/uBlockOrigin/uAssets/issues/22274#issuecomment-2745408699)).

The _only_ exception is if your path contains a comma, this will trip the parser, in which case the workaround is to simply use a regex instead, i.e. `example.com/,/after_comma##...` => `/example.com\/\x2C\/after_comma/##...`.

Also, it makes more sense to convert AG's `[$path=...]` syntax to the new support for path parts, converting to `:matches-path()` was working only for cosmetic filters, and this led to [such issue](https://github.com/uBlockOrigin/uBlock-issues/issues/3235).

**Reply by u-RraaLL** [reply]
Support for scriptlets and HTML filtering.

Possibly more efficient than the procedural for cosmetic filters too?
If not, then at least easier for newbies to use since people still regularly ask why `example.com/path##selector` filters don't work.

**Reply by u-RraaLL** [reply]
It doesn't work on reddit.

    reddit.com/r/uBlockOrigin/##main,#siteTable

The selectors work fine without the path (first one new, 2nd old).

**Reply by u-RraaLL** [reply]
~Also, I'm wondering if it would be useful for users if there was a ctrl+click or shift+click shortcut on the [Create] button in picker (and maybe DOM inspector for exceptions).~

~Or a checkbox~
~[x] Apply the filter current page only~
~to have it work on mobile too.~

~[x] Apply to current page instead of whole website~

---

Edit: On second thought, maybe it's not the best idea after all since it wouldn't apply to just the current page (unless uBO would automatically apply regex to make it so). And we probably shouldn't encourage people to use this over page-specific selectors either.

**Reply by gorhill** [reply]
> It doesn't work on reddit.

Silly mistake, will fix in next dev build. I originally tested cases where the hostname part is exactly the same as the filter one.

**Reply by u-RraaLL** [reply]
I see. I'm so used to removing `www.` that I haven't thought of that. Works fine now, thanks.

**Reply by Yuki2718** [reply]
Will `matches-path` be deprecated?

**Reply by gorhill** [reply]
No, `:matches-path` is still useful for dynamically updated pages.

**Reply by stephenhawk8054** [reply]
> No, `:matches-path` is still useful for dynamically updated pages.

I don't get yet. Can you give example contexts?

**Reply by gorhill** [reply]
I want to remove the right sidebar but only in uBlockOrigin sub, add:

    reddit.com/r/uBlockOrigin###right-sidebar-container

- Visit `https://www.reddit.com/r/uBlockOrigin/`
- Result: Works
- Click a link to visit another sub (I have `https://www.reddit.com/r/firefox/` proposed to me in the left sidebar)
- Site update dynamically
- Result: Right sidebar is still hidden = not what I want

Now with:

    reddit.com##:matches-path(/r/uBlockOrigin) #right-sidebar-container

Works as expected, because it's a procedural cosmetic filter and it reacts to URL changes despite no real navigation occurring used by dynamically updated pages.

**Reply by stephenhawk8054** [reply]
I see. I get it now, so this reacts and injects based on the sites' navigation load, like scriptlets.

**Reply by u-RraaLL** [reply]
When using the new path option, is there a drop in efficiency compared to plain old domains without path? And is there a difference when using a regex path instead (or are they all converted to regex internally)?

**Reply by gorhill** [reply]
I benchmarked and there was no difference -- actually I think there was slight (impossible to perceive for end users) improvement due to reworking the code responsible to lookup static extended filters. Typically static extended filters are looked up once per context load, and I measure 20µs-40µs per document load when benchmarking, so a complete non-issue performance wise.

These sort of filters will always be a minority (as per AG's `[$path=...]` usage) and as such I expect will never be an issue performance-wise. No regex is used when using `domain/path` form, only when using `/domain/` or `/domain\/path/` forms.

In summary, no performance impact, and per usual advice, use such filters when nothing simpler solves a filtering issue.

**Reply by u-RraaLL** [reply]
Ok, thanks. I was mainly thinking of custom filters here.

Last question. Would it be possible (or even make sense) to add some type of path limiter similar to pipe in network filtering to match only the exact path specified (e.g. `/r/uBlockOrigin`) without matching its subpages (e.g. `r/uBlockOrigin/wiki`). Without resorting to regex, I mean.

**Reply by gorhill** [reply]
I think it's best to fall back to using regex when there is such need. Using special characters just open the door to more issues when the special character is needed as literal in the path. I don't want to start to deal with special characters and special escape sequences.

**Reply by u-RraaLL** [reply]
Makes sense.

So efficiency-wise, is this the correct order for cosmetic filters?

1. `example.com##.pageSpecificSelector` / `example.com/path##.pageSpecificSelector`
2. `/example\.com\/path/##selector`
3. `example.com##:macthes-path(/path) selector`

**Reply by gorhill** [reply]
Yes.

**Reply by krystian3w** [reply]
Only me ~or all world~:  https://github.com/uBlockOrigin/uBlock-issues/issues/3235#issuecomment-2790614346 ❔ Seems only me due mix `*` aka all URL-s with `*::selection`.

**Reply by uBlock-user** [reply]
@gorhill Any plans to add path support for scriptlets ?

**Reply by u-RraaLL** [reply]
What do you mean? It's already implemented.

**Reply by uBlock-user** [reply]
> It's already implemented.

It doesn't work.

**Reply by u-RraaLL** [reply]
Oh, you're right. I've only tried if it worked on a specific path that I chose and not if other paths were excluded.

**Reply by gorhill** [reply]
It's theoretically supported but the uBO's caching of scriptlets gets in the way, hence why it works fine on uBOL, but not uBO. I know the solution for this, but it's not trivial few-line changes, and I have been waiting for actual real world cases in uBO lists as motivation -- there also many others things I want to address so for now it's not a priority.

---

Just to add, it's uBO's caching and Firefox's specific code which register scriptlets as content scripts, the API has hostname granularity, but anyways the solution I have in mind works for both cases.

---

My mistake, regex- or path-based matching is not yet supported in uBOL, I confused with ancestor-based matching, which works better in uBOL than uBO, again because of cache getting in the way in uBO. But then all regex-, path-, and ancestor-based matching have the same solution alluded above.

**Reply by uBlock-user** [reply]
> I have been waiting for actual real world cases in uBO lists as motivation -- there also many others things I want to address so for now it's not a priority.

Yeah I don't have one, but I wanted to know whether you're open to the can of worms that have been opened after adding support for cosmetic filtering and also the GH commit makes no mention of scriptlets there . Should add this to TO-DO list for future though.

**Reply by gorhill** [reply]
I am open to do my best to provide the needed features to people who spend a lot of their time maintaining the lists, I don't see this as a can of worms.
